File: 2022/rust/src/bin/day02.rs
use rust::read_input;

#[derive(PartialEq, Debug, Copy, Clone)]
enum Shape {
    Rock = 1,
    Paper,
    Scissors,
}

impl Shape {
    fn new(input: char) -> Shape {
        match input {
            'A' | 'X' => Shape::Rock,
            'B' | 'Y' => Shape::Paper,
            'C' | 'Z' => Shape::Scissors,
            _ => panic!("Unexpected input"),
        }
    }

    fn beats(&self, other: &Shape) -> bool {
        match self {
            Shape::Rock => other == &Shape::Scissors,
            Shape::Paper => other == &Shape::Rock,
            Shape::Scissors => other == &Shape::Paper,
        }
    }

    fn score_against(&self, other: &Shape) -> i32 {
        let value = *self as i32;
        if self == other {
            3 + value
        } else if self.beats(other) {
            6 + value
        } else {
            value
        }
    }
}

pub fn main() {
    let input = read_input("day02");
    let rounds = input.split("\n");

    // Part 1
    let score = rounds.clone().map(|round| {
        let opponent = Shape::new(round.chars().next().unwrap());
        let player = Shape::new(round.chars().nth(2).unwrap());
        player.score_against(&opponent)
    }).sum::<i32>();
    println!("Strategy 1 score: {:?}", score);


    // Part 2
    let score = rounds.map(|round| {
        let opponent = Shape::new(round.chars().next().unwrap()) as i32;
        match round.chars().nth(2).unwrap() {
            'X' => if opponent == 1 { 3 } else { opponent - 1 },
            'Y' => opponent + 3,
            'Z' => opponent % 3 + 7,
            _ => panic!("Unexpected input"),
        }
    }).sum::<i32>();
    println!("Strategy 2 score: {:?}", score);
}

File: 2022/typescript/day02.ts
import Library from "./lib.ts";

const getRounds = () => Library.getInput("day02").split("\n");

enum Shape {
  rock = 1,
  paper,
  scissors,
}

enum Outcome {
  win,
  draw,
  loss,
}

const shapeMap: { [key: string]: Shape } = {
  A: Shape.rock,
  B: Shape.paper,
  C: Shape.scissors,
  X: Shape.rock,
  Y: Shape.paper,
  Z: Shape.scissors,
};

const outcomeMap: { [key: string]: Outcome } = {
  X: Outcome.loss,
  Y: Outcome.draw,
  Z: Outcome.win,
};

const shapeForOutcomeMap = {
  [Shape.rock]: {
    [Outcome.win]: Shape.paper,
    [Outcome.draw]: Shape.rock,
    [Outcome.loss]: Shape.scissors,
  },
  [Shape.paper]: {
    [Outcome.win]: Shape.scissors,
    [Outcome.draw]: Shape.paper,
    [Outcome.loss]: Shape.rock,
  },
  [Shape.scissors]: {
    [Outcome.win]: Shape.rock,
    [Outcome.draw]: Shape.scissors,
    [Outcome.loss]: Shape.paper,
  },
};

const rockPaperScissors = (player: Shape, opponent: Shape): number => {
  let score = Number(player);

  if (player === opponent) {
    score += 3;
  } else if (player === shapeForOutcomeMap[opponent][Outcome.win]) {
    score += 6;
  }

  return score;
};

// Part 1
export const strategy1Score = (): number => {
  let score = 0;
  for (const round of getRounds()) {
    const [opponent, player] = round.split(" ");
    score += rockPaperScissors(shapeMap[player], shapeMap[opponent]);
  }
  return score;
};
console.log("Strategy 1 score:", strategy1Score());

// Part 2
export const strategy2Score = (): number => {
  let score = 0;
  for (const round of getRounds()) {
    const [input1, input2] = round.split(" ");
    const opponent = shapeMap[input1];
    const player = shapeForOutcomeMap[opponent][outcomeMap[input2]];
    score += rockPaperScissors(player, opponent);
  }
  return score;
};
console.log("Strategy 2 score:", strategy2Score());
